systemtest/src/test/java/io/strimzi/systemtest/upgrade/AbstractUpgradeST.java

```diff
@@ -108,12 +112,56 @@ protected static JsonArray readUpgradeJson(String fileName) {
             procedures.put("interBrokerProtocolVersion", testKafkaVersion.protocolVersion());
             data.put("proceduresAfterOperatorUpgrade", procedures);
 
-            parameters.add(Arguments.of(data.getString("fromVersion"), "HEAD", data));
+            // fromVersion is the mid step here, for CO upgrade we'll use this: 'prevVersion' -> 'fromVersion' -> HEAD
+            parameters.add(Arguments.of(data.getString("prevVersion"), data.getString("fromVersion"), "HEAD", data));
         });
 
         return parameters.stream();
     }
 
+    protected static Map<String, JsonObject> buildMidStepUpgradeData(JsonObject jsonData) {
+        List<TestKafkaVersion> testKafkaVersions = TestKafkaVersion.getKafkaVersions();
+        TestKafkaVersion testKafkaVersion = testKafkaVersions.get(testKafkaVersions.size() - 1);
+
+        Map<String, JsonObject> steps = new HashMap<>();
+
+        String midStepUrl = jsonData.getString("urlFrom");
+        String midStepVersion = jsonData.getString("fromVersion");
+        String midStepExamples = jsonData.getString("fromExamples");
+
+        JsonObject conversionTool = jsonData.getJsonObject("conversionTool");
+
+        // X -> 0.22.0 data
+        JsonObject midStep = JsonObject.mapFrom(jsonData);
+        midStep.put("urlFrom", jsonData.getString("urlPrevVersion"));
+        midStep.put("fromVersion", jsonData.getString("prevVersion"));
+        midStep.put("fromExamples", jsonData.getString("prevVersionExamples"));
+
+        midStep.put("urlTo", midStepUrl);
+        midStep.put("toVersion", midStepVersion);
+        midStep.put("toExamples", midStepExamples);
+        midStep.put("urlToConversionTool", conversionTool.getString("urlToConversionTool"));
+        midStep.put("toConversionTool", conversionTool.getString("toConversionTool"));
+
+        JsonObject midStepProcedures = new JsonObject();
+        midStepProcedures.put("kafkaVersion", testKafkaVersion.version());
+        midStepProcedures.put("logMessageVersion", testKafkaVersion.messageVersion());
```

Comment:
Would be great to have some abstraction class, which will stores all properties inside JSON so we can avoid typo errors.

Reply:
Yes, that's sounds good, let me do it in next PR